src/group_theory/pi_hom.lean

```diff
@@ -0,0 +1,392 @@
+/-
+Copyright (c) 2022 Jocchim Breitner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+import group_theory.general_commutator
+import group_theory.order_of_element
+import data.finset.noncomm_prod
+import ring_theory.coprime.lemmas
+
+/-!
+# Canonical homomorphism from a pi group
+
+This file defines the construction of the canoncial homomorphism from a product group.
+
+Given a family of morphisms `ϕ i : H i →* G` for each `i : I` where elements in the
+image of different morphism commute, we obtain a canoncial morphism
+`pi_hom.hom : (Π (i : I), H i) →* G` that coincides with `ϕ``
+
+## Main definitions
+
+* `pi_hom.hom : (Π (i : I), H i) →* G` is the main homomorphism
+* `pi_hom_restr.hom (S : fintype S): (Π (i : I), H i) →* G` is the homomorphism restricted to the
+   set `S`, and mostly of internal interest in this file, to allow inductive proofs.
+* `subgroup_pi_hom.hom : (Π (i : I), H i) →* G` is the specialization to `H i : subgroup G` and
+   the subgroup embedding.
+
+## Main theorems
+
+* `pi_hom.hom` conicides with each `ϕ`
+* `pi_hom.range`: The range of `pi_hom.hom` is `⨆ (i : I), (ϕ i).range`
+* The range of `subgroup_pi_hom.hom` is `⨆ (i : I), H i`
+* `pi_hom.pow`: `pi_hom.hom` commutes with potentation.
+* `pi_hom.injective_of_independent`: `pi_hom.hom` is injective if the `ϕ` are injective and the
+   ranges of the `ϕ` are independent.
+* `independent_range_of_coprime_order`: If the `ϕ` are injective and the `H i` have coprime orders,
+   then the ranges are independent.
+* `independent_of_coprime_order`: If commuting, normal `H i` have coprime orders, they are
+   independent.
+
+-/
+
+open_locale big_operators
+
+-- A lot of faff just to transport `is_coprime.prod_left` from ℤ to ℕ
+lemma coprime_prod_left
+  {I : Type*}
+  {x : ℕ} {s : I → ℕ} {t : finset I} :
+  (∀ (i : I), i ∈ t → nat.coprime (s i) x) → nat.coprime (∏ (i : I) in t, s i) x :=
+begin
+  intro h,
+  rw ← nat.is_coprime_iff_coprime,
+  have := @nat.cast_prod _ ℤ _ s t,
+  simp [ -nat.cast_prod ] at this,
+  rw this,
+  apply is_coprime.prod_left,
+  intros i hi,
+  rw nat.is_coprime_iff_coprime,
+  apply h i hi,
+end
+
+section with_group
+
+parameters {G : Type*} [group G]
+
+-- TODO: Move to suitable file
+@[simp]
+lemma order_of_inv (x : G) : order_of x⁻¹ = order_of x :=
+begin
+  apply nat.dvd_antisymm; rewrite order_of_dvd_iff_pow_eq_one,
+  { rw [inv_pow, pow_order_of_eq_one, one_inv], },
+  { nth_rewrite 0 ← (inv_inv x), rw [inv_pow, pow_order_of_eq_one, one_inv], }
+end
+
+-- TODO: Move to suitable file? Or too specialized?
+lemma mul_eq_one_of_disjoint
+  {H₁ H₂ : subgroup G} (hdis : disjoint H₁ H₂) {x y : G} (hx : x ∈ H₁) (hy : y ∈ H₂)
+  (heq : x * y = 1) : x = 1 ∧ y = 1 :=
+begin
```

Comment:
A possibly useful lemma for `disjoint` groups, but not that deep actually. Move there, keep here (with `private`), inline into its use below?